syntax = "proto3";

package killrvideo.user_management;
option csharp_namespace = "KillrVideo.UserManagement";

import "common/common_types.proto";

// The service responsible for managing user information
service UserManagementService {
  // Creates a new user
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
  
  // Verify a user's username and password
  rpc VerifyCredentials(VerifyCredentialsRequest) returns (VerifyCredentialsResponse);
  
  // Gets a user or group of user's profiles
  rpc GetUserProfile(GetUserProfileRequest) returns (GetUserProfileResponse);
}

// Request to create a new user
message CreateUserRequest {
  killrvideo.common.Uuid user_id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  string password = 5;
}

// Response when creating a new user
message CreateUserResponse {
}

// Request to verify a user's credentials (i.e. for logging them in)
message VerifyCredentialsRequest {
  string email = 1;
  string password = 2;
}

// Response that indicates the user's id if the credentials were correct
message VerifyCredentialsResponse {
  killrvideo.common.Uuid user_id = 1;
}

// Request to get a user or multiple users profiles
message GetUserProfileRequest {
  repeated killrvideo.common.Uuid user_ids = 1;
}

// Response with user profiles
message GetUserProfileResponse {
  repeated UserProfile profiles = 1;
}

// A user's profile information
message UserProfile {
  killrvideo.common.Uuid user_id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
}